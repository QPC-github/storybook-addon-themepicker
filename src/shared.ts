/*************************************************************
 Copyright 2018-2019 eBay Inc.
 Developer/Architect: Frederik Goris

 Use of this source code is governed by an MIT-style
 license that can be found in the LICENSE file or at
 https://opensource.org/licenses/MIT.
 ************************************************************/

export const MESSAGE_SET_THEME_LIST = 'ebay/themepicker/set-theme-list';
export const MESSAGE_SELECT_THEME = 'ebay/themepicker/select-theme';
