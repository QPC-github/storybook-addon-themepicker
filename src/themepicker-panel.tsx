/*************************************************************
 Copyright 2018-2019 eBay Inc.
 Developer/Architect: Frederik Goris

 Use of this source code is governed by an MIT-style
 license that can be found in the LICENSE file or at
 https://opensource.org/licenses/MIT.
 ************************************************************/

import {css} from 'emotion';
import * as React from 'react';
import {MESSAGE_SELECT_THEME, MESSAGE_SET_THEME_LIST} from './shared';
import {StorybookAPI} from './storybook__addons-more';
import * as Channel from '@storybook/channels';

type ThemePickerPanelProps = {
  channel: Channel;
  api: StorybookAPI;
};
type ThemePickerPanelState = {
  themeNames: string[];
};

export class ThemePickerPanel extends React.Component<ThemePickerPanelProps, ThemePickerPanelState> {
  stopListeningOnStory: () => void;

  constructor(props: ThemePickerPanelProps) {
    super(props);
    this.state = {
      themeNames: []
    };
  }

  setThemeNames = (themeNames: string[]) => this.setState({themeNames});

  componentDidMount() {
    this.props.channel.on(MESSAGE_SET_THEME_LIST, this.setThemeNames);
    this.stopListeningOnStory = this.props.api.onStory(() => this.setThemeNames([]));
  }

  componentWillUnmount() {
    if (this.stopListeningOnStory) {
      this.stopListeningOnStory();
    }
    this.props.channel.removeListener(MESSAGE_SET_THEME_LIST, this.setThemeNames);
  }

  selectTheme = (index: number) => this.props.channel.emit(MESSAGE_SELECT_THEME, index);

  render() {
    const themeNames = this.state.themeNames;
    if (themeNames.length === 0) {
      return null;
    }
    return (
      <div className={ThemePickerPanelStyles.container}>
        <label htmlFor="theme-picker" className={ThemePickerPanelStyles.label}>
          Theme
        </label>
        <select
          id="theme-picker"
          className={ThemePickerPanelStyles.select}
          onChange={e => this.selectTheme(parseInt(e.target.value, 10))}
        >
          {themeNames.map((themeName, index) => (
            <option key={index + themeName} value={index}>
              {themeName}
            </option>
          ))}
        </select>
      </div>
    );
  }
}

class ThemePickerPanelStyles {
  static container = css`
    padding: 15px;
    width: 100%;
    display: flex;
    font-family: 'Market Sans', 'Helvetica Neue', Helvetica, Arial, sans-serif;
    color: rgb(68, 68, 68);
  `;
  static label = css`
    width: 80px;
    margin-right: 15px;
    line-height: 30px;
  `;
  static select = css`
    color: rgb(85, 85, 85);
    width: 100%;
    border: 1px solid rgb(247, 247, 247);
    background-color: rgb(247, 247, 247);
    border-radius: 3px;
    height: 30px;
  `;
}
