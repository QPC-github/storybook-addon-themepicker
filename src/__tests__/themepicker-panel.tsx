import * as React from 'react';
import {mount} from 'enzyme';
import {ThemePickerPanel} from '../themepicker-panel';
import {mockChannel, mockStorybookAPI} from '../../test/mocks';
import {configure} from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import {MESSAGE_SELECT_THEME, MESSAGE_SET_THEME_LIST} from '../shared';
import * as EventEmitter from 'events';

configure({adapter: new Adapter()});

describe('ThemePickerPanel', () => {
  beforeEach(() => {
    const mockOn: jest.Mock<(type: PropertyKey, listener: (...args: any) => void) => void> = mockChannel.on as any;
    mockOn.mockClear();
    const mockEmit: jest.Mock<(type: PropertyKey, ...args: any) => void> = mockChannel.emit as any;
    mockEmit.mockClear();
    const mockOnStory: jest.Mock<(callback: (kind: string, story?: string) => void) => () => void> = mockStorybookAPI.onStory as any;
    mockOnStory.mockClear();
  });

  it('should mount', () => {
    const wrapper = mount(<ThemePickerPanel channel={mockChannel} api={mockStorybookAPI}/>);
    wrapper.setState({themeNames: ['Theme A', 'Theme B', 'Theme C']});
    expect(wrapper).toMatchSnapshot();
  });

  it('should display the theme names in a select menu', () => {
    const wrapper = mount(<ThemePickerPanel channel={mockChannel} api={mockStorybookAPI}/>);
    wrapper.setState({themeNames: ['Theme A', 'Theme B', 'Theme C']});
    expect(wrapper.find('select')).toHaveLength(1);
    expect(wrapper.find('option')).toHaveLength(3);
    expect(wrapper.find('option').at(0).text()).toEqual('Theme A');
    expect(wrapper.find('option').at(1).text()).toEqual('Theme B');
    expect(wrapper.find('option').at(2).text()).toEqual('Theme C');
  });

  it('should subscribe to MESSAGE_SET_THEME_LIST events, and set the theme names on that event', () => {
    const wrapper = mount(<ThemePickerPanel channel={mockChannel} api={mockStorybookAPI}/>);
    const mockOn: jest.Mock<any> = mockChannel.on as any;
    expect(mockOn).toHaveBeenCalledWith(MESSAGE_SET_THEME_LIST, expect.any(Function));
    let instance = wrapper.instance();
    expect(instance.state).toEqual({themeNames: []});
    const callback = mockOn.mock.calls[0][1];
    let themeNames = ['Theme 1', 'Theme 2', 'Theme 3'];
    callback(themeNames);
    expect(instance.state).toEqual({themeNames});
  });

  it('should send a MESSAGE_SELECT_THEME event when a user selects a theme', () => {
    const wrapper = mount(<ThemePickerPanel channel={mockChannel} api={mockStorybookAPI}/>);
    wrapper.setState({themeNames: ['Theme A', 'Theme B', 'Theme C']});
    const themeCValue = wrapper.find('option').at(2).prop('value');
    wrapper.find('select').simulate('change', {target: {value: themeCValue}});
    expect(mockChannel.emit).toHaveBeenCalledWith(MESSAGE_SELECT_THEME, 2);
  });

  it('should unsubscribe from MESSAGE_SET_THEME_LIST when it unmounts', () => {
    const eventEmitter = new EventEmitter();
    expect(eventEmitter.listenerCount(MESSAGE_SET_THEME_LIST)).toEqual(0);
    const wrapper = mount(<ThemePickerPanel channel={eventEmitter as any} api={mockStorybookAPI}/>);
    expect(eventEmitter.listenerCount(MESSAGE_SET_THEME_LIST)).toEqual(1);
    wrapper.unmount();
    expect(eventEmitter.listenerCount(MESSAGE_SET_THEME_LIST)).toEqual(0);
  });

  it('should reset the themes to an empty array when opening a new story', () => {
    const wrapper = mount(<ThemePickerPanel channel={mockChannel} api={mockStorybookAPI}/>);
    let mockOnStory: jest.Mock<(callback: (kind: string, story?: string) => void) => () => void> = mockStorybookAPI.onStory as any;
    expect(mockOnStory).toHaveBeenCalledWith(expect.any(Function));
    const callback = mockOnStory.mock.calls[0][0];
    wrapper.setState({themeNames: ['Default', 'Dark', 'Sparkly']});
    expect(wrapper.instance().state).toEqual({themeNames: ['Default', 'Dark', 'Sparkly']});
    callback('some.kind.of', 'Story');
    expect(wrapper.instance().state).toEqual({themeNames: []});
  });

  it('should not render unless it receives theme names', () => {
    const wrapper = mount(<ThemePickerPanel channel={mockChannel} api={mockStorybookAPI}/>);
    expect(wrapper.html()).toBeNull();
  });
});
