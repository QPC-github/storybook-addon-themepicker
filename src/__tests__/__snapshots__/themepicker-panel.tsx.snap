// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemePickerPanel should mount 1`] = `
.emotion-2 {
  padding: 15px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: 'Market Sans','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: rgb(68,68,68);
}

.emotion-0 {
  width: 80px;
  margin-right: 15px;
  line-height: 30px;
}

.emotion-1 {
  color: rgb(85,85,85);
  width: 100%;
  border: 1px solid rgb(247,247,247);
  background-color: rgb(247,247,247);
  border-radius: 3px;
  height: 30px;
}

<ThemePickerPanel
  api={
    Object {
      "onStory": [MockFunction] {
        "calls": Array [
          Array [
            [Function],
          ],
        ],
      },
    }
  }
  channel={
    Object {
      "addListener": [MockFunction],
      "addPeerListener": [MockFunction],
      "emit": [MockFunction],
      "eventNames": [MockFunction],
      "listenerCount": [MockFunction],
      "listeners": [MockFunction],
      "on": [MockFunction] {
        "calls": Array [
          Array [
            "ebay/themepicker/set-theme-list",
            [Function],
          ],
        ],
      },
      "once": [MockFunction],
      "prependListener": [MockFunction],
      "removeAllListeners": [MockFunction],
      "removeListener": [MockFunction],
    }
  }
>
  <div
    className="emotion-2"
  >
    <label
      className="emotion-0"
      htmlFor="theme-picker"
    >
      Theme
    </label>
    <select
      className="emotion-1"
      id="theme-picker"
      onChange={[Function]}
    >
      <option
        key="0Theme A"
        value={0}
      >
        Theme A
      </option>
      <option
        key="1Theme B"
        value={1}
      >
        Theme B
      </option>
      <option
        key="2Theme C"
        value={2}
      >
        Theme C
      </option>
    </select>
  </div>
</ThemePickerPanel>
`;
